lib/danger/request_source/bitbucket_server.rb

```diff
@@ -0,0 +1,113 @@
+
+module Danger
+  module RequestSources
+    class BitbucketServer < RequestSource
+      include Danger::Helpers::CommentsHelper
+      attr_accessor :pr_json
+
+      def initialize(ci_source, environment)
+         self.ci_source = ci_source
+         self.environment = environment
+
+         @username = @environment["DANGER_BITBUCKETSERVER_USERNAME"]
+         @password = @environment["DANGER_BITBUCKETSERVER_PASSWORD"]
+      end
+
+      def validates_as_ci?
+        # TODO ???
+        true
+      end
+      
+      def validates_as_api_source?
+        @username && !@username.empty? && @password && !@password.empty?
+      end
+
+      def scm
+        @scm ||= GitRepo.new
+      end
+
+      def host
+        @host ||= @environment["DANGER_BITBUCKETSERVER_HOST"]
+      end
+
+      def pr_api_endpoint
+        project, slug = ci_source.repo_slug.split("/")
+        "https://#{host}/rest/api/1.0/projects/#{project}/repos/#{slug}/pull-requests/#{ci_source.pull_request_id}"
+      end
+
+      def fetch_details
+        self.pr_json = get_json(URI(pr_api_endpoint))
+      end
+
+      def setup_danger_branches
+        base_commit = self.pr_json[:toRef][:latestCommit]
+        head_commit = self.pr_json[:fromRef][:latestCommit]
+
+        # Next, we want to ensure that we have a version of the current branch at a known location
+        self.scm.exec "branch #{EnvironmentManager.danger_base_branch} #{base_commit}"
+
+        # OK, so we want to ensure that we have a known head branch, this will always represent
+        # the head of the PR ( e.g. the most recent commit that will be merged. )
+        self.scm.exec "branch #{EnvironmentManager.danger_head_branch} #{head_commit}"
+      end
+
+      def organisation
+        nil
+      end
+      
+      def update_pull_request!(warnings: [], errors: [], messages: [], markdowns: [], danger_id: 'danger')
+        delete_old_comments(danger_id: danger_id)
+        
+        comment = generate_description(warnings: warnings, errors: errors)
+        comment += "\n\n"
+        comment += generate_comment(warnings: warnings,
+                                     errors: errors,
+                                   messages: messages,
+                                  markdowns: markdowns,
+                        previous_violations: {},
+                                  danger_id: danger_id,
+                                   template: "bitbucket_server")
+
+        post(URI("#{pr_api_endpoint}/comments"), {text: comment}.to_json)
+      end
+
+      def delete_old_comments(danger_id: 'danger')
+        uri = URI("#{pr_api_endpoint}/activities?limit=1000")
+        comments = get_json(uri)[:values]
+          .select { |v| v[:action] == "COMMENTED" }
+          .map { |v| v[:comment] }
+          .select { |c| c[:text] =~ /generated_by_#{danger_id}/ }
+          .each { |c| delete(URI("#{pr_api_endpoint}/comments/#{c[:id]}?version=#{c[:version]}")) }
+      end
+
+      def get_json(uri)
```

Comment:
Might wanna move the logic around talking to the API to a BitBucketServerClient. Violates SRP a bit now when it's doing both the low level HTTP calls and the high level danger stuff